### PR TITLE
BLE central Role implementation and peripheral refactor

### PR DESCRIPTION
**Pull request by sgbihu**
@calvinatintel @SidLeung @eriknyquist 

I'm sorry for I make the old pull request lost when try to reset my local branch. The logs are lost too.

### CODE REVIEW

**Comment by eriknyquist** [comment]
@sgbihu please read our [commit message guidelines](https://github.com/01org/corelibs-arduino101/wiki/Writing-a-commit-message) and follow them. You have 3 commits here; one of them adds a single line to a file, and the very next one removes that same line. You also added a README change, which is unrelated to BLE, to your BLE commit.

Can we please have 2 commits; one for the BLE central implementation, and one for the README update.

File: README.md [hunk 1]
```diff
@@ -53,3 +53,21 @@ them to the [support forum](https://forum.arduino.cc/index.php?board=103).
 > "How do I use this library?"
 
 > "I can't get this example sketch to work. What am I doing wrong?"
+
+# Enable debug interface on Serail1
```
Comment (eriknyquist) [inline README.md]: Should be Serial1

**Comment by eriknyquist** [comment]
I'm sorry @sgbihu, I misspoke; the commit message on your large BLE commit is actually fine. So the only things that need to be done here are 1) remove the first two commits, and 2) split each task in the big BLE commit into separate commits. We can still review it, in the meantime, but if you can fix these things before it gets merged we would appreciate it.

File: cores/arduino/printk.cpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdarg.h>
+#include "UARTClass.h"
+
+extern "C" void printk(const char *fmt, va_list args);
```
Comment (eriknyquist) [inline cores/arduino/printk.cpp]: Is printk really needed for BLE central?

File: cores/arduino/printk.cpp [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+						}
+						while (pos >= 9) {
+							if (found_largest_digit
+							    || remainder >
+							    pos) {
+								found_largest_digit
+									= 1;
+								if (len < size)
+									*dest++
+										=
```
Comment (eriknyquist) [inline cores/arduino/printk.cpp]: Looks like the formatting got a little messed up for the next 10 lines here, can you clean it up please?

File: cores/arduino/printk.cpp [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+	int len = 0;
+
+    char tmp[PRINTK_BUFSIZ];
+
+    len = __vsnprintf(tmp, PRINTK_BUFSIZ, fmt, args);
+    
+    tmp[len] = '\0';
+    Serial1.println(tmp);
+#endif
+}
```
Comment (eriknyquist) [inline cores/arduino/printk.cpp]: "cores/arduino/wiring_shift.c" this file has a permission mode change, and nothing else, committed. Can you remove it please? (Sorry, I couldn't leave a comment on the permission change itself, so I'm leaving it here on an unrelated file)

File: libraries/CurieBLE/src/BLECentral.cpp [hunk 5]
```diff
@@ -1,103 +1,82 @@
-/*
```
Comment (eriknyquist) [inline libraries/CurieBLE/src/BLECentral.cpp]: How come the license was removed from this file?

File: libraries/CurieBLE/examples/IMUBleCentral/IMUBleCentral.ino [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+unsigned seqNum = 0;
+
+long previousMillis = 0;  // last time the battery level was checked, in ms
+
+BLECentral bleCentral;;       // BLE Central Device (the board you're programming)
+
+BLEService bleImuService("F7580001-153E-D4F6-F26D-43D8D98EEB13");
+// Tx IMU data Characteristic
+BLECharacteristic bleImuChar("F7580003-153E-D4F6-F26D-43D8D98EEB13",       // standard 16-bit characteristic UUID
```
Comment (eriknyquist) [inline libraries/CurieBLE/examples/IMUBleCentral/IMUBleCentral.ino]: formatting and comments is very messy here, can you clean it up please

File: libraries/CurieBLE/examples/IMUBleCentral/IMUBleCentral.ino [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+void setup() {
+    Serial1.begin(115200);
```
Comment (eriknyquist) [inline libraries/CurieBLE/examples/IMUBleCentral/IMUBleCentral.ino]: Why are both Serial and Serial1 needed for this example sketch? can't you just print all the information to one serial port?

File: libraries/CurieBLE/examples/IMUBleNotification/IMUBleNotification.ino [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+BLEPeripheral blePeripheral;       // BLE Peripheral Device (the board you're programming)
+BLEService bleImuService("F7580001-153E-D4F6-F26D-43D8D98EEB13");
+
+// Tx IMU data Characteristic
+
+BLECharacteristic bleImuChar("F7580003-153E-D4F6-F26D-43D8D98EEB13",       // standard 16-bit characteristic UUID
+                             BLERead | BLENotify, sizeof(imuBuf)); 	   // remote clients will be able to
+			     	       		  			   // get notifications if this characteristic changes
+void setup() {
+    Serial1.begin(115200); // For corelib debug
```
Comment (eriknyquist) [inline libraries/CurieBLE/examples/IMUBleNotification/IMUBleNotification.ino]: Serial1 not used, remove this please

File: libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+void setup() {
+    Serial1.begin(115200);
```
Comment (eriknyquist) [inline libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino]: Serial1 not used, remove please

File: libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    // assign event handlers for characteristic
+    switchChar.setEventHandler(BLEWritten, switchCharacteristicWritten);
+
+    bleCentral.begin();
+    Serial.println(("Bluetooth device active, waiting for connections..."));
+}
+
+void loop()
+{
+    static unsigned char counter = 0;
```
Comment (eriknyquist) [inline libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino]: Why do you need `static unsigned char` and `static char` for these? `counter` is ... well, just a counter, and `ledstate` only flips back and forth between 0 and 1. To avoid confusing users, maybe use `int` or `short` for both?

File: libraries/CurieBLE/examples/Scanning/Scanning.ino [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    }
+    adv_list_add(device);
+    return true;
+}
+
+void setup() {
+    log_init();
+    Serial1.begin(115200);
+
+    Serial.begin(115200);    // initialize serial communication
```
Comment (eriknyquist) [inline libraries/CurieBLE/examples/Scanning/Scanning.ino]: Please use Serial, not Serial1, for printing information to the user.

File: libraries/CurieBLE/src/BLECentral.h [hunk 12]
```diff
@@ -1,70 +1,121 @@
-/*
```
Comment (eriknyquist) [inline libraries/CurieBLE/src/BLECentral.h]: LIcense has been removed

File: libraries/CurieBLE/src/BLECentralRole.cpp [hunk 13]
```diff
@@ -0,0 +1,274 @@
+
```
Comment (eriknyquist) [inline libraries/CurieBLE/src/BLECentralRole.cpp]: No license in this file either- please add licenses to all the source files that don't already have them

File: libraries/CurieBLE/src/BLEProfile.cpp [hunk 14]
```diff
@@ -0,0 +1,513 @@
...
+    }
+    if (NULL == _attr_base)
+    {
+        _attr_base = (struct bt_gatt_attr *)malloc((BLEAttribute::numAttributes() + BLECharacteristic::numNotifyAttributes()) * sizeof(struct bt_gatt_attr));
+        memset(_attr_base, 0x00, ((BLEAttribute::numAttributes() + BLECharacteristic::numNotifyAttributes()) * sizeof(struct bt_gatt_attr)));
+        pr_info(LOG_MODULE_BLE, "_attr_base_-%p, size-%d", _attr_base, sizeof(_attr_base));
+    }
+    if (NULL == _sub_param)
+    {
+        _sub_param = (struct bt_gatt_subscribe_params *)malloc((BLECharacteristic::numNotifyAttributes()) * sizeof(struct bt_gatt_subscribe_params));
```
Comment (eriknyquist) [inline libraries/CurieBLE/src/BLEProfile.cpp]: I see that `_attributes` and `_attr_base` are freed in the `BLEProfile` class destructor, but I don't see anything in the library that frees `_sub_param`. Will this memory be lost after BLE Central is used?

File: libraries/CurieBLE/src/BLEProfile.cpp [hunk 15]
```diff
@@ -0,0 +1,513 @@
...
+// Only for central
+uint8_t profile_notify_process (struct bt_conn *conn,
+                             struct bt_gatt_subscribe_params *params,
+                             const void *data, uint16_t length)
+{
+    BLEPeripheralHelper* peripheral = BLECentralRole::instance()->peripheral(conn);// Find peripheral by bt_conn
+    BLEAttribute* notifyatt = peripheral->attribute(params); // Find attribute by params
+    BLECharacteristic *chrc = (BLECharacteristic *)notifyatt;
+    
+    //assert(notifyatt->type() == BLETypeCharacteristic);
```
Comment (eriknyquist) [inline libraries/CurieBLE/src/BLEProfile.cpp]: Commented assert-- left over debug code?

File: libraries/CurieBLE/src/BLEProfile.cpp [hunk 16]
```diff
@@ -0,0 +1,513 @@
...
+
+
+void BLEProfile::clearHandles(void)
+{
+    int i;
+    struct bt_gatt_attr *attr = NULL;
+    // Didn't need to unsubscribe
+    //  The stack will unsubscribe the notify when disconnected.
+    //  The sub_param has some pointer. So can't call memset. Just reset the index.
+    //memset(_sub_param, 0x00, ((BLECharacteristic::numNotifyAttributes()) * sizeof(struct bt_gatt_subscribe_params)));
```
Comment (eriknyquist) [inline libraries/CurieBLE/src/BLEProfile.cpp]: Is this commented memset call part of a TODO for future work? If so, maybe just put an extra note specifying this. If not, can you remove it?

File: libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino (hunk 10 above)
Comment (eriknyquist) [inline libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino]: Actually I see why you need static, so that the values of `counter` and `ledstate` persist between `loop()` calls. And the `ledstate` is `unsigned char` because that's what `BLECharacteristic::write()` accepts. Could you still use an `int` or `short` for `counter`?

File: cores/arduino/printk.cpp (hunk 2 above)
Comment (sgbihu) [inline cores/arduino/printk.cpp]: Not for central. Only for debug.

File: README.md (hunk 1 above)
Comment (sgbihu) [inline README.md]: Change the typo

File: cores/arduino/printk.cpp (hunk 3 above)
Comment (sgbihu) [inline cores/arduino/printk.cpp]: Change the formate

File: cores/arduino/printk.cpp (hunk 4 above)
Comment (sgbihu) [inline cores/arduino/printk.cpp]: I didn't change this file. It should be referenced from merge or rebase. I will fix this.

File: libraries/CurieBLE/src/BLECentral.cpp (hunk 5 above)
Comment (sgbihu) [inline libraries/CurieBLE/src/BLECentral.cpp]: Add license in all new file. This is a new class.

File: libraries/CurieBLE/examples/IMUBleCentral/IMUBleCentral.ino (hunk 6 above)
Comment (sgbihu) [inline libraries/CurieBLE/examples/IMUBleCentral/IMUBleCentral.ino]: Modify the comments and formatting. Delete some unused variable

File: libraries/CurieBLE/examples/IMUBleCentral/IMUBleCentral.ino (hunk 7 above)
Comment (sgbihu) [inline libraries/CurieBLE/examples/IMUBleCentral/IMUBleCentral.ino]: This is for serial debug via Serial1. Will delete it in example sketches.

File: libraries/CurieBLE/examples/IMUBleNotification/IMUBleNotification.ino (hunk 8 above)
Comment (sgbihu) [inline libraries/CurieBLE/examples/IMUBleNotification/IMUBleNotification.ino]: Delete all Serial1 in example sketches.

File: libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino (hunk 9 above)
Comment (sgbihu) [inline libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino]: Same as previous.

File: libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino (hunk 10 above)
Comment (sgbihu) [inline libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino]: Change counter as unsigned int. For the counter can't be negative number.

File: libraries/CurieBLE/examples/Scanning/Scanning.ino (hunk 11 above)
Comment (sgbihu) [inline libraries/CurieBLE/examples/Scanning/Scanning.ino]: Modified.

File: libraries/CurieBLE/src/BLECentral.h (hunk 12 above)
Comment (sgbihu) [inline libraries/CurieBLE/src/BLECentral.h]: Add license.

File: libraries/CurieBLE/src/BLECentralRole.cpp (hunk 13 above)
Comment (sgbihu) [inline libraries/CurieBLE/src/BLECentralRole.cpp]: Add license.

File: libraries/CurieBLE/src/BLEProfile.cpp (hunk 14 above)
Comment (sgbihu) [inline libraries/CurieBLE/src/BLEProfile.cpp]: Fix this bug.

File: libraries/CurieBLE/src/BLEProfile.cpp (hunk 15 above)
Comment (sgbihu) [inline libraries/CurieBLE/src/BLEProfile.cpp]: For the project didn't define the assert. Will uncomment it after we define the interface. And this is in my plan. Try to let function works first.

File: libraries/CurieBLE/src/BLEProfile.cpp (hunk 16 above)
Comment (sgbihu) [inline libraries/CurieBLE/src/BLEProfile.cpp]: Remove the line 459

File: cores/arduino/printk.cpp (hunk 2 above)
Comment (eriknyquist) [inline cores/arduino/printk.cpp]: I guess I'm just not sure why we need a new function to print formatted strings, when there are already some options for that available. Which specific features of vsnprintf are required to debug BLE Central?

File: libraries/CurieBLE/src/BLECentralRole.cpp [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+
+BLECentralRole* BLECentralRole::_ble_central_ins = NULL;
+
+BLECentralRole *BLECentralRole::instance()
+{
+    if (NULL == _ble_central_ins)
+    {
+        _ble_central_ins = new BLECentralRole();
```
Comment (eriknyquist) [inline libraries/CurieBLE/src/BLECentralRole.cpp]: I think `_ble_central_ins` should also be deleted in `BLECentralRole::~BLECentralRole`, otherwise that memory will be lost after using BLE

File: libraries/CurieBLE/src/BLEPeripheralRole.cpp [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+    _max_conn_interval(DEFAULT_MAX_CONN_INTERVAL),
+    _peripheral(NULL),
+    _central(this)
+{
+    memset(_event_handlers, 0x00, sizeof(_event_handlers));
+    _peripheral.setAddress(_local_bda);
+}
+
+BLEPeripheralRole::~BLEPeripheralRole(void)
+{
```
Comment (eriknyquist) [inline libraries/CurieBLE/src/BLEPeripheralRole.cpp]: `delete(_ins);` ?

File: cores/arduino/printk.cpp (hunk 2 above)
Comment (sgbihu) [inline cores/arduino/printk.cpp]: Remove this function and use vsnprintf in cstdio.

Comment (sgbihu) [inline cores/arduino/printk.cpp]: Just want debug corelib easier.

File: libraries/CurieBLE/src/BLECentralRole.cpp (hunk 17 above)
Comment (sgbihu) [inline libraries/CurieBLE/src/BLECentralRole.cpp]: The BLECentralRole and BLEPeripheralRole is a hardware abstract layer for application. So the instance will always exist when start BLE. And this variable is static. No memory leak.

File: libraries/CurieBLE/src/BLEPeripheralRole.cpp (hunk 18 above)
Comment (sgbihu) [inline libraries/CurieBLE/src/BLEPeripheralRole.cpp]: Same as previous. Move this topic to previous one.

File: libraries/CurieBLE/src/BLECentralRole.cpp (hunk 17 above)
Comment (eriknyquist) [inline libraries/CurieBLE/src/BLECentralRole.cpp]: `static BLECentralRole* _ble_central_ins;`
Well, the pointer is static, but not what it points to. At the end of the day, malloc gets called when you use the `new` keyword (see the override in `cores/arduino/new.cpp`).

File: libraries/CurieBLE/examples/Scanning/Scanning.ino [hunk 19]
```diff
@@ -0,0 +1,129 @@
+/*
```
Comment (eriknyquist) [inline libraries/CurieBLE/examples/Scanning/Scanning.ino]: Can you please put some explanatory comments at the top of each example sketch? i.e. what the sketch does and maybe a quick summary of how it works. Look at the existing built-in example sketches for reference.

File: libraries/CurieBLE/src/BLECentralRole.cpp (hunk 17 above)
Comment (sgbihu) [inline libraries/CurieBLE/src/BLECentralRole.cpp]: Yes, it call the malloc and get the memory resource. But the stack and Nordic’s code didn’t have API to deregister the BLE profile. And the stack also use the pointer that point to the CurieBLE LIB. If we release those resource, the stack will have wild pointer and we didn't have a API to clean it. So we can't release those resource. Our current design use singleton pattern. Those resource can be managed and never leak.

Comment (sgbihu) [inline libraries/CurieBLE/src/BLECentralRole.cpp]: We also can release those resource. But the BLE can't call begin after call end. I can change the code if you think this is acceptable.

**Comment by noelpaz** [comment]
There is a runtime issue where a peripheral cannot be discovered or not advertising . Please see https://github.com/01org/corelibs-arduino101/issues/258

File: libraries/CurieBLE/examples/IMUBleNotification/IMUBleNotification.ino [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#include <CurieBLE.h>
+#include <CurieIMU.h>
+
+/*
+   This sketch example partially implements the standard Bluetooth Low-Energy Battery service.
```
Comment (eriknyquist) [inline libraries/CurieBLE/examples/IMUBleNotification/IMUBleNotification.ino]: Is this comment accurate?

File: libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-
+  1301 USA
+*/
+
+// This example can work with CallbackLED to show the profile read/write operation in central
```
Comment (eriknyquist) [inline libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino]: Are CallbackLED and LEDCentral required to work together? if so, this should be more obvious. Both sketches should have an explanation at the top, describing how the two sketches should be used together, and they should be named similarly, e.g. "LEDCentral" and "LEDPeripheral" or something like that.

File: libraries/CurieBLE/examples/IMUBleCentral/IMUBleCentral.ino [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#include <CurieBLE.h>
+
+/*
+   This sketch example partially implements the standard Bluetooth Low-Energy Battery service.
```
Comment (eriknyquist) [inline libraries/CurieBLE/examples/IMUBleCentral/IMUBleCentral.ino]: Can you remove this comment and replace it with something that describes what this sketch does?

File: libraries/CurieBLE/examples/IMUBleNotification/IMUBleNotification.ino (hunk 20 above)
Comment (sgbihu) [inline libraries/CurieBLE/examples/IMUBleNotification/IMUBleNotification.ino]: Change the comments.

File: libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino (hunk 21 above)
Comment (sgbihu) [inline libraries/CurieBLE/examples/LEDCentral/LEDCentral.ino]: I have tested this example with CallbackLED. I think every sketches has a goal. I have designed those examples, Scan, establish connection, read/write profile and process notification. For the peripheral, I think we have some APP to test it. Do you think we need align the peripheral and central?

File: libraries/CurieBLE/examples/IMUBleCentral/IMUBleCentral.ino (hunk 22 above)
Comment (sgbihu) [inline libraries/CurieBLE/examples/IMUBleCentral/IMUBleCentral.ino]: Change the comments.

File: libraries/CurieBLE/examples/IMUBleNotification/IMUBleNotification.ino (hunk 20 above)
Comment (eriknyquist) [inline libraries/CurieBLE/examples/IMUBleNotification/IMUBleNotification.ino]: It looks the same... I'm talking about this line
`This sketch example partially implements the standard Bluetooth Low-Energy Battery service`

It's the same line from the BatteryMonitor sketch. Can you replace it with something that explains to the user what the sketch is doing? and likewise for the other examples?

File: libraries/CurieBLE/examples/Scanning/Scanning.ino (hunk 19 above)
Comment (sgbihu) [inline libraries/CurieBLE/examples/Scanning/Scanning.ino]: Add some comment to describe the function.

**Comment by calvinatintel** [comment]
`LED` and `LED Central` sketches need some work. While `LED Central` is more robust, `LED` often fails to run, or even put the board in a state which requires a manual reset. I've had to upload `LED` multiple times and reboot the board multiple times in order to get it to pair with a device running `LED Central`. Once paired, I can reset the `LED Central` board repeatedly and the connection will resume as soon as the board boots back up. However, resetting the device running `LED` will break the pairing, and the `LED` running device doesn't recover until it's reflashed and reset many times.

**Comment by noelpaz** [comment]
@calvinatintel .  Have you tried using another serial terminal for the LED  sketch and then push the LEDCentral after it is running. I find running 2 instances of the IDE for Serial Monitor is problematic. Anyway I will test this again.

**Comment by noelpaz** [comment]
I have not counted but I have also noticed having to hit the reset button more than once and it is not exclusive to the BLE sketches. I am using 1.6.11

**Comment by sgbihu** [comment]
Hi @calvinatintel and @noelpaz ,

Can we move to Jira? I create a Jira tickets. https://jira.ndg.intel.com/browse/ATLEDGE-680

**Comment by noelpaz** [comment]
That's okay -- I saw the fix for JIRA-680

File: libraries/CurieBLE/src/BLEPeripheralRole.cpp [hunk 23]
```diff
@@ -295,7 +295,7 @@ void BLEPeripheralRole::handleParamUpdated(struct bt_conn *conn,
                         conn, interval, latency, timeout);
     if (_event_handlers[BLEUpdateParam])
     {
-        _central.setConnParames(interval, interval, latency, timeout);
+        _central.setConnectionParameters(interval, interval, latency, timeout);
```
Comment (noelpaz) [inline libraries/CurieBLE/src/BLEPeripheralRole.cpp]: @sgbihu -- do we have a test sketch that will test these changes

**Comment by noelpaz** [comment]
@sgbihu -- For JIRA 685 Janet wanted a cooment on BatteryAdvChange example. On the top of the cooment. Please add

"This sketch is not paired with a specific central example sketch, but to see how it works you need to use a BLE app on your phone or central device and try connecting when it is either a connectable or not connectable state as displayed in the serial monitor."

**Comment by calvinatintel** [comment]
Merging in!
